File: bindings/nodejs/lib/ufsecp.js
/**
 * UltrafastSecp256k1 — Node.js FFI binding (ufsecp stable C ABI v1).
 *
 * High-performance secp256k1 elliptic curve cryptography with dual-layer
 * constant-time architecture. Context-based API.
 *
 * Usage:
 *   const { Ufsecp } = require('ufsecp');
 *   const ctx = new Ufsecp();
 *   const pub = ctx.pubkeyCreate(Buffer.alloc(32, 0).fill(1, 31));
 *   ctx.destroy();
 */

'use strict';

const ffi = require('ffi-napi');
const ref = require('ref-napi');
const path = require('path');
const os = require('os');

const voidPtr = ref.refType(ref.types.void);
const voidPtrPtr = ref.refType(voidPtr);
const uint8Ptr = ref.refType(ref.types.uint8);
const int32Ptr = ref.refType(ref.types.int32);
const sizeTPtr = ref.refType(ref.types.size_t);

// ── Error codes ────────────────────────────────────────────────────────

const UFSECP_OK              = 0;
const UFSECP_ERR_NULL_ARG    = 1;
const UFSECP_ERR_BAD_KEY     = 2;
const UFSECP_ERR_BAD_PUBKEY  = 3;
const UFSECP_ERR_BAD_SIG     = 4;
const UFSECP_ERR_BAD_INPUT   = 5;
const UFSECP_ERR_VERIFY_FAIL = 6;
const UFSECP_ERR_ARITH       = 7;
const UFSECP_ERR_SELFTEST    = 8;
const UFSECP_ERR_INTERNAL    = 9;
const UFSECP_ERR_BUF_SMALL   = 10;

const ERROR_NAMES = {
  [UFSECP_ERR_NULL_ARG]:    'null argument',
  [UFSECP_ERR_BAD_KEY]:     'invalid private key',
  [UFSECP_ERR_BAD_PUBKEY]:  'invalid public key',
  [UFSECP_ERR_BAD_SIG]:     'invalid signature',
  [UFSECP_ERR_BAD_INPUT]:   'bad input',
  [UFSECP_ERR_VERIFY_FAIL]: 'verification failed',
  [UFSECP_ERR_ARITH]:       'arithmetic error',
  [UFSECP_ERR_SELFTEST]:    'selftest failed',
  [UFSECP_ERR_INTERNAL]:    'internal error',
  [UFSECP_ERR_BUF_SMALL]:   'buffer too small',
};

class UfsecpError extends Error {
  constructor(op, code) {
    super(`ufsecp ${op} failed: ${ERROR_NAMES[code] || `unknown (${code})`}`);
    this.name = 'UfsecpError';
    this.code = code;
    this.operation = op;
  }
}

// ── Library resolution ─────────────────────────────────────────────────

function findLibrary() {
  const plat = os.platform();
  const name = plat === 'win32' ? 'ufsecp.dll'
    : plat === 'darwin' ? 'libufsecp.dylib'
    : 'libufsecp.so';

  // 1. UFSECP_LIB env var
  const env = process.env.UFSECP_LIB;
  if (env) return env;

  // 2. Next to this file
  const here = path.join(__dirname, name);
  try { require('fs').accessSync(here); return here; } catch {}

  // 3. prebuilds/<platform>-<arch>/
  const prebuilt = path.join(__dirname, '..', 'prebuilds',
    `${plat}-${os.arch()}`, name);
  try { require('fs').accessSync(prebuilt); return prebuilt; } catch {}

  // 4. System default
  return name;
}

// ── FFI binding ────────────────────────────────────────────────────────

const LIB_SPEC = {
  // Context
  ufsecp_ctx_create:    ['int', [voidPtrPtr]],
  ufsecp_ctx_destroy:   ['void', [voidPtr]],
  ufsecp_ctx_clone:     ['int', [voidPtr, voidPtrPtr]],
  // Version
  ufsecp_version:       ['uint32', []],
  ufsecp_abi_version:   ['uint32', []],
  ufsecp_version_string:['string', []],
  ufsecp_error_str:     ['string', ['int']],
  ufsecp_last_error:    ['int', [voidPtr]],
  ufsecp_last_error_msg:['string', [voidPtr]],
  // Key ops
  ufsecp_pubkey_create:             ['int', [voidPtr, uint8Ptr, uint8Ptr]],
  ufsecp_pubkey_create_uncompressed:['int', [voidPtr, uint8Ptr, uint8Ptr]],
  ufsecp_pubkey_parse:              ['int', [voidPtr, uint8Ptr, 'size_t', uint8Ptr]],
  ufsecp_pubkey_xonly:              ['int', [voidPtr, uint8Ptr, uint8Ptr]],
  ufsecp_seckey_verify:             ['int', [voidPtr, uint8Ptr]],
  ufsecp_seckey_negate:             ['int', [voidPtr, uint8Ptr]],
  ufsecp_seckey_tweak_add:          ['int', [voidPtr, uint8Ptr, uint8Ptr]],
  ufsecp_seckey_tweak_mul:          ['int', [voidPtr, uint8Ptr, uint8Ptr]],
  // ECDSA
  ufsecp_ecdsa_sign:                ['int', [voidPtr, uint8Ptr, uint8Ptr, uint8Ptr]],
  ufsecp_ecdsa_verify:              ['int', [voidPtr, uint8Ptr, uint8Ptr, uint8Ptr]],
  ufsecp_ecdsa_sig_to_der:          ['int', [voidPtr, uint8Ptr, uint8Ptr, sizeTPtr]],
  ufsecp_ecdsa_sig_from_der:        ['int', [voidPtr, uint8Ptr, 'size_t', uint8Ptr]],
  // Recovery
  ufsecp_ecdsa_sign_recoverable:    ['int', [voidPtr, uint8Ptr, uint8Ptr, uint8Ptr, int32Ptr]],
  ufsecp_ecdsa_recover:             ['int', [voidPtr, uint8Ptr, uint8Ptr, 'int', uint8Ptr]],
  // Schnorr
  ufsecp_schnorr_sign:              ['int', [voidPtr, uint8Ptr, uint8Ptr, uint8Ptr, uint8Ptr]],
  ufsecp_schnorr_verify:            ['int', [voidPtr, uint8Ptr, uint8Ptr, uint8Ptr]],
  // ECDH
  ufsecp_ecdh:                      ['int', [voidPtr, uint8Ptr, uint8Ptr, uint8Ptr]],
  ufsecp_ecdh_xonly:                ['int', [voidPtr, uint8Ptr, uint8Ptr, uint8Ptr]],
  ufsecp_ecdh_raw:                  ['int', [voidPtr, uint8Ptr, uint8Ptr, uint8Ptr]],
  // Hashing
  ufsecp_sha256:                    ['int', [uint8Ptr, 'size_t', uint8Ptr]],
  ufsecp_hash160:                   ['int', [uint8Ptr, 'size_t', uint8Ptr]],
  ufsecp_tagged_hash:               ['int', ['string', uint8Ptr, 'size_t', uint8Ptr]],
  // Addresses
  ufsecp_addr_p2pkh:                ['int', [voidPtr, uint8Ptr, 'int', uint8Ptr, sizeTPtr]],
  ufsecp_addr_p2wpkh:               ['int', [voidPtr, uint8Ptr, 'int', uint8Ptr, sizeTPtr]],
  ufsecp_addr_p2tr:                 ['int', [voidPtr, uint8Ptr, 'int', uint8Ptr, sizeTPtr]],
  // WIF
  ufsecp_wif_encode:                ['int', [voidPtr, uint8Ptr, 'int', 'int', uint8Ptr, sizeTPtr]],
  ufsecp_wif_decode:                ['int', [voidPtr, 'string', uint8Ptr, int32Ptr, int32Ptr]],
  // BIP-32
  ufsecp_bip32_master:              ['int', [voidPtr, uint8Ptr, 'size_t', uint8Ptr]],
  ufsecp_bip32_derive:              ['int', [voidPtr, uint8Ptr, 'uint32', uint8Ptr]],
  ufsecp_bip32_derive_path:         ['int', [voidPtr, uint8Ptr, 'string', uint8Ptr]],
  ufsecp_bip32_privkey:             ['int', [voidPtr, uint8Ptr, uint8Ptr]],
  ufsecp_bip32_pubkey:              ['int', [voidPtr, uint8Ptr, uint8Ptr]],
  // Taproot
  ufsecp_taproot_output_key:        ['int', [voidPtr, uint8Ptr, uint8Ptr, uint8Ptr, int32Ptr]],
  ufsecp_taproot_tweak_seckey:      ['int', [voidPtr, uint8Ptr, uint8Ptr, uint8Ptr]],
  ufsecp_taproot_verify:            ['int', [voidPtr, uint8Ptr, 'int', uint8Ptr, uint8Ptr, 'size_t']],
};

// ── Constants ──────────────────────────────────────────────────────────

const NET_MAINNET = 0;
const NET_TESTNET = 1;

// ── Ufsecp class ───────────────────────────────────────────────────────

class Ufsecp {
  /**
   * @param {string} [libPath] Path to the ufsecp shared library.
   */
  constructor(libPath) {
    this._lib = ffi.Library(libPath || findLibrary(), LIB_SPEC);
    const pp = ref.alloc(voidPtr);
    const rc = this._lib.ufsecp_ctx_create(pp);
    if (rc !== UFSECP_OK) throw new UfsecpError('ctx_create', rc);
    this._ctx = pp.deref();
    this._destroyed = false;
  }

  /** Explicitly destroy the context. Safe to call multiple times. */
  destroy() {
    if (!this._destroyed && this._ctx) {
      this._lib.ufsecp_ctx_destroy(this._ctx);
      this._ctx = null;
      this._destroyed = true;
    }
  }

  // ── Version ──────────────────────────────────────────────────────────

  version()       { return this._lib.ufsecp_version(); }
  abiVersion()    { return this._lib.ufsecp_abi_version(); }
  versionString() { return this._lib.ufsecp_version_string(); }
  lastError()     { this._alive(); return this._lib.ufsecp_last_error(this._ctx); }
  lastErrorMsg()  { this._alive(); return this._lib.ufsecp_last_error_msg(this._ctx); }

  // ── Key operations ────────────────────────────────────────────────────

  /** Compressed public key (33 bytes). */
  pubkeyCreate(privkey) {
    _chk(privkey, 32, 'privkey'); this._alive();
    const out = Buffer.alloc(33);
    this._throw(this._lib.ufsecp_pubkey_create(this._ctx, privkey, out), 'pubkey_create');
    return out;
  }

  /** Uncompressed public key (65 bytes). */
  pubkeyCreateUncompressed(privkey) {
    _chk(privkey, 32, 'privkey'); this._alive();
    const out = Buffer.alloc(65);
    this._throw(this._lib.ufsecp_pubkey_create_uncompressed(this._ctx, privkey, out), 'pubkey_create_uncompressed');
    return out;
  }

  /** Parse compressed/uncompressed → compressed 33 bytes. */
  pubkeyParse(pubkey) {
    this._alive();
    const out = Buffer.alloc(33);
    this._throw(this._lib.ufsecp_pubkey_parse(this._ctx, pubkey, pubkey.length, out), 'pubkey_parse');
    return out;
  }

  /** X-only (32 bytes, BIP-340) from private key. */
  pubkeyXonly(privkey) {
    _chk(privkey, 32, 'privkey'); this._alive();
    const out = Buffer.alloc(32);
    this._throw(this._lib.ufsecp_pubkey_xonly(this._ctx, privkey, out), 'pubkey_xonly');
    return out;
  }

  seckeyVerify(privkey)  { _chk(privkey, 32, 'privkey'); this._alive(); return this._lib.ufsecp_seckey_verify(this._ctx, privkey) === UFSECP_OK; }

  seckeyNegate(privkey) {
    _chk(privkey, 32, 'privkey'); this._alive();
    const buf = Buffer.from(privkey);
    this._throw(this._lib.ufsecp_seckey_negate(this._ctx, buf), 'seckey_negate');
    return buf;
  }

  seckeyTweakAdd(privkey, tweak) {
    _chk(privkey, 32, 'privkey'); _chk(tweak, 32, 'tweak'); this._alive();
    const buf = Buffer.from(privkey);
    this._throw(this._lib.ufsecp_seckey_tweak_add(this._ctx, buf, tweak), 'seckey_tweak_add');
    return buf;
  }

  seckeyTweakMul(privkey, tweak) {
    _chk(privkey, 32, 'privkey'); _chk(tweak, 32, 'tweak'); this._alive();
    const buf = Buffer.from(privkey);
    this._throw(this._lib.ufsecp_seckey_tweak_mul(this._ctx, buf, tweak), 'seckey_tweak_mul');
    return buf;
  }

  // ── ECDSA ─────────────────────────────────────────────────────────────

  ecdsaSign(msgHash, privkey) {
    _chk(msgHash, 32, 'msgHash'); _chk(privkey, 32, 'privkey'); this._alive();
    const sig = Buffer.alloc(64);
    this._throw(this._lib.ufsecp_ecdsa_sign(this._ctx, msgHash, privkey, sig), 'ecdsa_sign');
    return sig;
  }

  ecdsaVerify(msgHash, sig, pubkey) {
    _chk(msgHash, 32, 'msgHash'); _chk(sig, 64, 'sig'); _chk(pubkey, 33, 'pubkey'); this._alive();
    return this._lib.ufsecp_ecdsa_verify(this._ctx, msgHash, sig, pubkey) === UFSECP_OK;
  }

  ecdsaSigToDer(sig) {
    _chk(sig, 64, 'sig'); this._alive();
    const der = Buffer.alloc(72);
    const len = ref.alloc('size_t', 72);
    this._throw(this._lib.ufsecp_ecdsa_sig_to_der(this._ctx, sig, der, len), 'ecdsa_sig_to_der');
    return der.subarray(0, len.deref());
  }

  ecdsaSigFromDer(der) {
    this._alive();
    const sig = Buffer.alloc(64);
    this._throw(this._lib.ufsecp_ecdsa_sig_from_der(this._ctx, der, der.length, sig), 'ecdsa_sig_from_der');
    return sig;
  }

  // ── Recovery ──────────────────────────────────────────────────────────

  ecdsaSignRecoverable(msgHash, privkey) {
    _chk(msgHash, 32, 'msgHash'); _chk(privkey, 32, 'privkey'); this._alive();
    const sig = Buffer.alloc(64);
    const recid = ref.alloc('int');
    this._throw(this._lib.ufsecp_ecdsa_sign_recoverable(this._ctx, msgHash, privkey, sig, recid), 'ecdsa_sign_recoverable');
    return { signature: sig, recoveryId: recid.deref() };
  }

  ecdsaRecover(msgHash, sig, recid) {
    _chk(msgHash, 32, 'msgHash'); _chk(sig, 64, 'sig'); this._alive();
    const pub = Buffer.alloc(33);
    this._throw(this._lib.ufsecp_ecdsa_recover(this._ctx, msgHash, sig, recid, pub), 'ecdsa_recover');
    return pub;
  }

  // ── Schnorr ───────────────────────────────────────────────────────────

  schnorrSign(msg, privkey, auxRand) {
    _chk(msg, 32, 'msg'); _chk(privkey, 32, 'privkey'); _chk(auxRand, 32, 'auxRand'); this._alive();
    const sig = Buffer.alloc(64);
    this._throw(this._lib.ufsecp_schnorr_sign(this._ctx, msg, privkey, auxRand, sig), 'schnorr_sign');
    return sig;
  }

  schnorrVerify(msg, sig, pubkeyX) {
    _chk(msg, 32, 'msg'); _chk(sig, 64, 'sig'); _chk(pubkeyX, 32, 'pubkeyX'); this._alive();
    return this._lib.ufsecp_schnorr_verify(this._ctx, msg, sig, pubkeyX) === UFSECP_OK;
  }

  // ── ECDH ──────────────────────────────────────────────────────────────

  ecdh(privkey, pubkey) {
    _chk(privkey, 32, 'privkey'); _chk(pubkey, 33, 'pubkey'); this._alive();
    const out = Buffer.alloc(32);
    this._throw(this._lib.ufsecp_ecdh(this._ctx, privkey, pubkey, out), 'ecdh');
    return out;
  }

  ecdhXonly(privkey, pubkey) {
    _chk(privkey, 32, 'privkey'); _chk(pubkey, 33, 'pubkey'); this._alive();
    const out = Buffer.alloc(32);
    this._throw(this._lib.ufsecp_ecdh_xonly(this._ctx, privkey, pubkey, out), 'ecdh_xonly');
    return out;
  }

  ecdhRaw(privkey, pubkey) {
    _chk(privkey, 32, 'privkey'); _chk(pubkey, 33, 'pubkey'); this._alive();
    const out = Buffer.alloc(32);
    this._throw(this._lib.ufsecp_ecdh_raw(this._ctx, privkey, pubkey, out), 'ecdh_raw');
    return out;
  }

  // ── Hashing ───────────────────────────────────────────────────────────

  sha256(data) {
    const out = Buffer.alloc(32);
    this._throw(this._lib.ufsecp_sha256(data, data.length, out), 'sha256');
    return out;
  }

  hash160(data) {
    const out = Buffer.alloc(20);
    this._throw(this._lib.ufsecp_hash160(data, data.length, out), 'hash160');
    return out;
  }

  taggedHash(tag, data) {
    const out = Buffer.alloc(32);
    this._throw(this._lib.ufsecp_tagged_hash(tag, data, data.length, out), 'tagged_hash');
    return out;
  }

  // ── Addresses ─────────────────────────────────────────────────────────

  addrP2PKH(pubkey, network = NET_MAINNET) {
    _chk(pubkey, 33, 'pubkey'); return this._getAddr('ufsecp_addr_p2pkh', pubkey, network);
  }

  addrP2WPKH(pubkey, network = NET_MAINNET) {
    _chk(pubkey, 33, 'pubkey'); return this._getAddr('ufsecp_addr_p2wpkh', pubkey, network);
  }

  addrP2TR(xonlyKey, network = NET_MAINNET) {
    _chk(xonlyKey, 32, 'xonlyKey'); return this._getAddr('ufsecp_addr_p2tr', xonlyKey, network);
  }

  // ── WIF ───────────────────────────────────────────────────────────────

  wifEncode(privkey, compressed = true, network = NET_MAINNET) {
    _chk(privkey, 32, 'privkey'); this._alive();
    const buf = Buffer.alloc(128);
    const len = ref.alloc('size_t', 128);
    this._throw(this._lib.ufsecp_wif_encode(this._ctx, privkey, compressed ? 1 : 0, network, buf, len), 'wif_encode');
    return buf.toString('utf8', 0, len.deref());
  }

  wifDecode(wif) {
    this._alive();
    const key = Buffer.alloc(32);
    const comp = ref.alloc('int');
    const net = ref.alloc('int');
    this._throw(this._lib.ufsecp_wif_decode(this._ctx, wif, key, comp, net), 'wif_decode');
    return { privkey: key, compressed: comp.deref() === 1, network: net.deref() };
  }

  // ── BIP-32 ────────────────────────────────────────────────────────────

  bip32Master(seed) {
    this._alive();
    if (seed.length < 16 || seed.length > 64) throw new RangeError('Seed must be 16-64 bytes');
    const key = Buffer.alloc(82);
    this._throw(this._lib.ufsecp_bip32_master(this._ctx, seed, seed.length, key), 'bip32_master');
    return key;
  }

  bip32Derive(parent, index) {
    _chk(parent, 82, 'parent'); this._alive();
    const child = Buffer.alloc(82);
    this._throw(this._lib.ufsecp_bip32_derive(this._ctx, parent, index >>> 0, child), 'bip32_derive');
    return child;
  }

  bip32DerivePath(master, path) {
    _chk(master, 82, 'master'); this._alive();
    const key = Buffer.alloc(82);
    this._throw(this._lib.ufsecp_bip32_derive_path(this._ctx, master, path, key), 'bip32_derive_path');
    return key;
  }

  bip32Privkey(key) {
    _chk(key, 82, 'key'); this._alive();
    const priv = Buffer.alloc(32);
    this._throw(this._lib.ufsecp_bip32_privkey(this._ctx, key, priv), 'bip32_privkey');
    return priv;
  }

  bip32Pubkey(key) {
    _chk(key, 82, 'key'); this._alive();
    const pub = Buffer.alloc(33);
    this._throw(this._lib.ufsecp_bip32_pubkey(this._ctx, key, pub), 'bip32_pubkey');
    return pub;
  }

  // ── Taproot ───────────────────────────────────────────────────────────

  taprootOutputKey(internalKeyX, merkleRoot = null) {
    _chk(internalKeyX, 32, 'internalKeyX'); this._alive();
    const out = Buffer.alloc(32);
    const parity = ref.alloc('int');
    this._throw(this._lib.ufsecp_taproot_output_key(this._ctx, internalKeyX, merkleRoot, out, parity), 'taproot_output_key');
    return { outputKeyX: out, parity: parity.deref() };
  }

  taprootTweakSeckey(privkey, merkleRoot = null) {
    _chk(privkey, 32, 'privkey'); this._alive();
    const out = Buffer.alloc(32);
    this._throw(this._lib.ufsecp_taproot_tweak_seckey(this._ctx, privkey, merkleRoot, out), 'taproot_tweak_seckey');
    return out;
  }

  taprootVerify(outputKeyX, parity, internalKeyX, merkleRoot = null) {
    _chk(outputKeyX, 32, 'outputKeyX'); _chk(internalKeyX, 32, 'internalKeyX'); this._alive();
    const mrLen = merkleRoot ? merkleRoot.length : 0;
    return this._lib.ufsecp_taproot_verify(this._ctx, outputKeyX, parity, internalKeyX, merkleRoot, mrLen) === UFSECP_OK;
  }

  // ── Internal ──────────────────────────────────────────────────────────

  _alive() {
    if (this._destroyed) throw new Error('UfsecpContext already destroyed');
  }

  _throw(rc, op) {
    if (rc !== UFSECP_OK) throw new UfsecpError(op, rc);
  }

  _getAddr(fnName, key, network) {
    this._alive();
    const buf = Buffer.alloc(128);
    const len = ref.alloc('size_t', 128);
    this._throw(this._lib[fnName](this._ctx, key, network, buf, len), 'address');
    return buf.toString('utf8', 0, len.deref());
  }
}

function _chk(buf, expected, name) {
  if (!Buffer.isBuffer(buf) && !(buf instanceof Uint8Array)) {
    throw new TypeError(`${name} must be a Buffer`);
  }
  if (buf.length !== expected) {
    throw new RangeError(`${name} must be ${expected} bytes, got ${buf.length}`);
  }
}

module.exports = { Ufsecp, UfsecpError, NET_MAINNET, NET_TESTNET };

File: bindings/swift/Sources/Ufsecp/Ufsecp.swift
/// UltrafastSecp256k1 — Swift binding (ufsecp stable C ABI v1).
///
/// High-performance secp256k1 ECC with dual-layer constant-time architecture.
/// Context-based API.
///
/// Usage:
///     let ctx = try UfsecpContext()
///     let pubkey = try ctx.pubkeyCreate(privkey: Data(repeating: 0, count: 31) + Data([0x01]))
///     ctx.destroy()

import Foundation
#if canImport(CUfsecp)
import CUfsecp
#endif

// MARK: - Error

public enum UfsecpErrorCode: Int32 {
    case ok = 0
    case nullArg = 1
    case badKey = 2
    case badPubkey = 3
    case badSig = 4
    case badInput = 5
    case verifyFail = 6
    case arith = 7
    case selftest = 8
    case `internal` = 9
    case bufTooSmall = 10
}

public struct UfsecpError: Error, CustomStringConvertible {
    public let operation: String
    public let code: UfsecpErrorCode

    public var description: String {
        "ufsecp \(operation) failed: \(code) (\(code.rawValue))"
    }
}

// MARK: - Result types

public enum Network: Int32 {
    case mainnet = 0
    case testnet = 1
}

public struct RecoverableSignature {
    public let signature: Data
    public let recoveryId: Int32
}

public struct TaprootOutputKeyResult {
    public let outputKeyX: Data
    public let parity: Int32
}

public struct WifDecoded {
    public let privkey: Data
    public let compressed: Bool
    public let network: Network
}

// MARK: - Context

public final class UfsecpContext {
    private var ctx: OpaquePointer?
    private var destroyed = false

    public init() throws {
        var ptr: OpaquePointer?
        let rc = ufsecp_ctx_create(&ptr)
        guard rc == 0, let p = ptr else {
            throw UfsecpError(operation: "ctx_create", code: UfsecpErrorCode(rawValue: rc) ?? .internal)
        }
        self.ctx = p
    }

    deinit { destroy() }

    public func destroy() {
        guard !destroyed, let c = ctx else { return }
        ufsecp_ctx_destroy(c)
        ctx = nil
        destroyed = true
    }

    // MARK: Version

    public static var version: UInt32 { ufsecp_version() }
    public static var abiVersion: UInt32 { ufsecp_abi_version() }
    public static var versionString: String { String(cString: ufsecp_version_string()) }

    public var lastError: Int32 { try! alive(); return ufsecp_last_error(ctx!) }
    public var lastErrorMsg: String { try! alive(); return String(cString: ufsecp_last_error_msg(ctx!)) }

    // MARK: Key Operations

    public func pubkeyCreate(privkey: Data) throws -> Data {
        try chk(privkey, 32, "privkey"); try alive()
        var out = [UInt8](repeating: 0, count: 33)
        try privkey.withUnsafeBytes { pk in
            try throwRC(ufsecp_pubkey_create(ctx!, pk.baseAddress!.assumingMemoryBound(to: UInt8.self), &out), "pubkey_create")
        }
        return Data(out)
    }

    public func pubkeyCreateUncompressed(privkey: Data) throws -> Data {
        try chk(privkey, 32, "privkey"); try alive()
        var out = [UInt8](repeating: 0, count: 65)
        try privkey.withUnsafeBytes { pk in
            try throwRC(ufsecp_pubkey_create_uncompressed(ctx!, pk.baseAddress!.assumingMemoryBound(to: UInt8.self), &out), "pubkey_create_uncompressed")
        }
        return Data(out)
    }

    public func pubkeyParse(pubkey: Data) throws -> Data {
        try alive()
        var out = [UInt8](repeating: 0, count: 33)
        try pubkey.withUnsafeBytes { pk in
            try throwRC(ufsecp_pubkey_parse(ctx!, pk.baseAddress!.assumingMemoryBound(to: UInt8.self), pubkey.count, &out), "pubkey_parse")
        }
        return Data(out)
    }

    public func pubkeyXonly(privkey: Data) throws -> Data {
        try chk(privkey, 32, "privkey"); try alive()
        var out = [UInt8](repeating: 0, count: 32)
        try privkey.withUnsafeBytes { pk in
            try throwRC(ufsecp_pubkey_xonly(ctx!, pk.baseAddress!.assumingMemoryBound(to: UInt8.self), &out), "pubkey_xonly")
        }
        return Data(out)
    }

    public func seckeyVerify(privkey: Data) throws -> Bool {
        try chk(privkey, 32, "privkey"); try alive()
        return privkey.withUnsafeBytes { pk in
            ufsecp_seckey_verify(ctx!, pk.baseAddress!.assumingMemoryBound(to: UInt8.self)) == 0
        }
    }

    public func seckeyNegate(privkey: Data) throws -> Data {
        try chk(privkey, 32, "privkey"); try alive()
        var buf = [UInt8](privkey)
        try throwRC(ufsecp_seckey_negate(ctx!, &buf), "seckey_negate")
        return Data(buf)
    }

    public func seckeyTweakAdd(privkey: Data, tweak: Data) throws -> Data {
        try chk(privkey, 32, "privkey"); try chk(tweak, 32, "tweak"); try alive()
        var buf = [UInt8](privkey)
        try tweak.withUnsafeBytes { tw in
            try throwRC(ufsecp_seckey_tweak_add(ctx!, &buf, tw.baseAddress!.assumingMemoryBound(to: UInt8.self)), "seckey_tweak_add")
        }
        return Data(buf)
    }

    public func seckeyTweakMul(privkey: Data, tweak: Data) throws -> Data {
        try chk(privkey, 32, "privkey"); try chk(tweak, 32, "tweak"); try alive()
        var buf = [UInt8](privkey)
        try tweak.withUnsafeBytes { tw in
            try throwRC(ufsecp_seckey_tweak_mul(ctx!, &buf, tw.baseAddress!.assumingMemoryBound(to: UInt8.self)), "seckey_tweak_mul")
        }
        return Data(buf)
    }

    // MARK: ECDSA

    public func ecdsaSign(msgHash: Data, privkey: Data) throws -> Data {
        try chk(msgHash, 32, "msgHash"); try chk(privkey, 32, "privkey"); try alive()
        var sig = [UInt8](repeating: 0, count: 64)
        try msgHash.withUnsafeBytes { msg in
            try privkey.withUnsafeBytes { pk in
                try throwRC(ufsecp_ecdsa_sign(ctx!,
                    msg.baseAddress!.assumingMemoryBound(to: UInt8.self),
                    pk.baseAddress!.assumingMemoryBound(to: UInt8.self), &sig), "ecdsa_sign")
            }
        }
        return Data(sig)
    }

    public func ecdsaVerify(msgHash: Data, sig: Data, pubkey: Data) throws -> Bool {
        try chk(msgHash, 32, "msgHash"); try chk(sig, 64, "sig"); try chk(pubkey, 33, "pubkey"); try alive()
        return msgHash.withUnsafeBytes { msg in
            sig.withUnsafeBytes { s in
                pubkey.withUnsafeBytes { pk in
                    ufsecp_ecdsa_verify(ctx!,
                        msg.baseAddress!.assumingMemoryBound(to: UInt8.self),
                        s.baseAddress!.assumingMemoryBound(to: UInt8.self),
                        pk.baseAddress!.assumingMemoryBound(to: UInt8.self)) == 0
                }
            }
        }
    }

    // MARK: Schnorr

    public func schnorrSign(msg: Data, privkey: Data, auxRand: Data) throws -> Data {
        try chk(msg, 32, "msg"); try chk(privkey, 32, "privkey"); try chk(auxRand, 32, "auxRand"); try alive()
        var sig = [UInt8](repeating: 0, count: 64)
        try msg.withUnsafeBytes { m in
            try privkey.withUnsafeBytes { pk in
                try auxRand.withUnsafeBytes { ar in
                    try throwRC(ufsecp_schnorr_sign(ctx!,
                        m.baseAddress!.assumingMemoryBound(to: UInt8.self),
                        pk.baseAddress!.assumingMemoryBound(to: UInt8.self),
                        ar.baseAddress!.assumingMemoryBound(to: UInt8.self), &sig), "schnorr_sign")
                }
            }
        }
        return Data(sig)
    }

    public func schnorrVerify(msg: Data, sig: Data, pubkeyX: Data) throws -> Bool {
        try chk(msg, 32, "msg"); try chk(sig, 64, "sig"); try chk(pubkeyX, 32, "pubkeyX"); try alive()
        return msg.withUnsafeBytes { m in
            sig.withUnsafeBytes { s in
                pubkeyX.withUnsafeBytes { pk in
                    ufsecp_schnorr_verify(ctx!,
                        m.baseAddress!.assumingMemoryBound(to: UInt8.self),
                        s.baseAddress!.assumingMemoryBound(to: UInt8.self),
                        pk.baseAddress!.assumingMemoryBound(to: UInt8.self)) == 0
                }
            }
        }
    }

    // MARK: ECDH

    public func ecdh(privkey: Data, pubkey: Data) throws -> Data {
        try chk(privkey, 32, "privkey"); try chk(pubkey, 33, "pubkey"); try alive()
        var out = [UInt8](repeating: 0, count: 32)
        try privkey.withUnsafeBytes { pk in
            try pubkey.withUnsafeBytes { pub in
                try throwRC(ufsecp_ecdh(ctx!,
                    pk.baseAddress!.assumingMemoryBound(to: UInt8.self),
                    pub.baseAddress!.assumingMemoryBound(to: UInt8.self), &out), "ecdh")
            }
        }
        return Data(out)
    }

    // MARK: Hashing

    public static func sha256(_ data: Data) throws -> Data {
        var out = [UInt8](repeating: 0, count: 32)
        try data.withUnsafeBytes { d in
            try throwRC(ufsecp_sha256(d.baseAddress!.assumingMemoryBound(to: UInt8.self), data.count, &out), "sha256")
        }
        return Data(out)
    }

    public static func hash160(_ data: Data) throws -> Data {
        var out = [UInt8](repeating: 0, count: 20)
        try data.withUnsafeBytes { d in
            try throwRC(ufsecp_hash160(d.baseAddress!.assumingMemoryBound(to: UInt8.self), data.count, &out), "hash160")
        }
        return Data(out)
    }

    // MARK: Internal

    private func alive() throws {
        guard !destroyed else { throw UfsecpError(operation: "alive", code: .internal) }
    }

    private func chk(_ data: Data, _ expected: Int, _ name: String) throws {
        guard data.count == expected else {
            throw UfsecpError(operation: "\(name) size", code: .badInput)
        }
    }

    private static func throwRC(_ rc: Int32, _ op: String) throws {
        guard rc == 0 else {
            throw UfsecpError(operation: op, code: UfsecpErrorCode(rawValue: rc) ?? .internal)
        }
    }

    private func throwRC(_ rc: Int32, _ op: String) throws {
        guard rc == 0 else {
            throw UfsecpError(operation: op, code: UfsecpErrorCode(rawValue: rc) ?? .internal)
        }
    }
}
